### PR TITLE
[Compiler] Stateless OpalCompiler

### PR DESCRIPTION
**Pull request by dionisiydk**
1. PR makes OpalCompiler stateless (with temp state deprecation):

With the recent changes to the compiler it is now clear that we can easily remove #source and #ast variables and reuse the compiler instance for parsing/compilation of multiple sources. This way we can get rid of CompilationContext moving all its state to the compiler.  So that we will have logical reference from the method/AST to the producer compiler instance (currently it is the reference to the CompilationContext). Then many tools can be simplified by reusing a same compiler instance for own needs (highlighting for example). 

 Here PR performs main changes to the users of #compiler: 

   - Variables #source and #ast are removed (source is temporally deprecated for integration). 
   - API is changed to be #compile:/parse:/#evaluate: with source code as a parameter:

```Smalltalk
OpalCompiler new noPattern: true; compile: '1+2'.
```
Versus old style:
```Smalltalk
OpalCompiler new source: '1+2'; noPattern: true; compile.
```
The new style API was always there but it is now implemented without instance variables #source and #ast. 
Most of the changes in PR is such a rewrite from old style to new one.

   - #source:, #compile, #evaluate methods are deprecated (will be removed after NewTools and Spec fixes).
   - All users of compiler are adopted accordingly

The deprecation of #source variable is required to get safe independent adoption of NewTools and Spec. After that the variables will be removed.

The inline of CompilerContext will be done in separate PR.

2. Environment parameter is moved from CompilationContext into OCSemanticScope as #targetEnvironment which uses the environment of #targetClass by default. 

   - it makes the #environment configuration really working. Previously it did nothing. See the changes in #testCompileWithEnvironment. 

3. Introduce compilation option #optionEmbeddAST to remove #compileDoItFromAST method. Now #doIt compilation simply enables #optionEmbeddAST option (just as #optionEmbeddSources option) and a scope indirection is not needed anymore. See #generateMethod.

### CODE REVIEW

File: src/AST-Core/CompiledMethod.extension.st [hunk 1]
```diff
@@ -25,10 +25,9 @@ CompiledMethod >> parseTree [
 	"returns an AST for this method, do not cache it. (see #ast for the cached alternative)"
 	| ast |
 	ast := self methodClass compiler 
-		source: self sourceCode;
 		failBlock: [^ self decompile ];
 		class: self methodClass;
-		parse.
+		parse: self sourceCode.
```
Comment (dionisiydk) [inline src/AST-Core/CompiledMethod.extension.st:30]: Such move of #source: argument is the most of the changes here to remove the state from the compiler

File: src/OpalCompiler-Core/Behavior.extension.st [hunk 2]
```diff
@@ -70,6 +68,14 @@ Behavior >> evaluate: aString [
 	^self compiler evaluate: aString
 ]
 
+{ #category : #'*OpalCompiler-Core' }
+Behavior >> lookupLocalVar: aName [
+	^ self classLayout 
+		resolveSlot: aName asSymbol 
+		ifFound: [:var | var] 
+		ifNone: [ self innerBindingOf: aName ]
```
Comment (dionisiydk) [inline src/OpalCompiler-Core/Behavior.extension.st:76]: The local lookup is introduced to find any variable defined in the class without the full environment lookup. It is required to allow the override of an #environment of a class during a compilation. 

We should really review and unify all current lookup methods: #lookupVar:#lookupLocalVar:/#bindingOf:/#innerBindingOf:.

File: src/OpalCompiler-Core/CompilationContext.class.st [hunk 3]
```diff
@@ -120,6 +119,16 @@ CompilationContext class >> optionCleanBlockClosure: aBoolean [
 	^ self writeDefaultOption: #optionCleanBlockClosure value: aBoolean
 ]
 
+{ #category : #'options - settings API' }
+CompilationContext class >> optionEmbeddAST [
+	^ self readDefaultOption: #optionEmbeddAST
+]
```
Comment (dionisiydk) [inline src/OpalCompiler-Core/CompilationContext.class.st:135]: Here is new option #optionEmbeddAST

File: src/OpalCompiler-Core/CompilationContext.class.st [hunk 4]
```diff
@@ -488,7 +498,7 @@ CompilationContext >> noPattern: anObject [
 	anObject ifTrue: [
 		semanticScope := semanticScope asDoItScope.
 		"when we compile an expression, embedd sources, as the resulting method will never be installed"
-		self parseOptions: #(+ #optionEmbeddSources).
+		self parseOptions: #(+ #optionEmbeddSources + #optionEmbeddAST).
```
Comment (dionisiydk) [inline src/OpalCompiler-Core/CompilationContext.class.st:522]: And #optionEmbeddAST is now default option for doIts

File: src/OpalCompiler-Core/RBMethodNode.extension.st [hunk 5]
```diff
@@ -85,11 +85,14 @@ RBMethodNode >> generateIR [
...
 RBMethodNode >> generateMethod [
-	| methodTrailer |
+	| methodTrailer newMethod |
 	methodTrailer := self compilationContext optionEmbeddSources
 			ifTrue: [ CompiledMethodTrailer new sourceCode: source ]
 			ifFalse: [self compilationContext compiledMethodTrailer ].
-	^self generate: methodTrailer
+	newMethod := self generate: methodTrailer.
+	self compilationContext optionEmbeddAST ifTrue: [ 
+		 newMethod propertyAt: #ast put: self ].
```
Comment (dionisiydk) [inline src/OpalCompiler-Core/RBMethodNode.extension.st:100]: And this is where optionEmbeddAST is used

File: src/OpalCompiler-Tests/OCEnvironmentScopeTest.class.st [hunk 6]
```diff
@@ -10,22 +10,26 @@ Class {
 { #category : #'tests - environment' }
 OCEnvironmentScopeTest >> testCompileWithEnvironment [
 	"testing compiling with environments other then Smalltalk globals. Note: the class in the environment has to point to the environment. And we have to compile for a class in the environment"
-
-
 	| environment method return | 
 	environment := SystemDictionary new.
 	environment at: #MyClass put: Point copy.
-	(environment at: #MyClass) environment: environment.
+	environment at: #MyClass2 put: Association.
```
Comment (dionisiydk) [inline src/OpalCompiler-Tests/OCEnvironmentScopeTest.class.st:16]: This test demonstrates the environment override. In the current image it does not work.

**Comment by MarcusDenker** [comment]
failing tests not related
- #skipOnPharoCITestingEnvironment seems to not work on WIN and MAC #11425
- Failing on CI (random): CoNarrowHistoryFetcherTest>>testNarrowingReturnsSameElementsThatCallingDirectly #11368

**Comment by guillep** [comment]
What's the next step here?
I see many merge conflicts.

**Comment by dionisiydk** [comment]
> What's the next step here? I see many merge conflicts.

I am waiting @MarcusDenker review for this idea.

Conflicts are resolved

**Comment by dionisiydk** [comment]
conflicts were badly fixed . I will revisit it. But it does not prevent the review

**Comment by MarcusDenker** [comment]
The PR https://github.com/pharo-project/pharo/pull/13152 now adds some more state.. (see discussion)

 I need to think about this more (stateless compiler, CompilationContext is for sure now holding on to too much state...)

**Comment by privat** [comment]
I added the cleaning of the state in CompilationContext and OpalCompiler to #12883, I will think about that.

**Comment by Ducasse** [comment]
So what is the status? I have the impression that the issue is not going to be integratable.

**Comment by tesonep** [comment]
Stale PR

**Pull request closed without merging**